### ISSUE TITLE
For large sets of data the downloaded survey results are incomplete

### ISSUE THREAD

**Issue by MgowanoJr** [unicef/iogt#1309]
**To replicate:**

1. Go to South Africa Site
2. Navigate to Form Data
3. Open the Survey COVID-19 Health Worker Vaccine Survey (https://za.goodinternet.org/admin/forms/submissions/292/)
4. Download CSV of results

**Expectation:**

The dataset has 1180 Pages of 20 items which translates to a download with around 23,000 responses.

**Actual:**

The csv file download has around 4,800 responses only

**Comment by cbunicef** [comment]
A few notes to add on how this issue behaves:

- When downloading XLSX file a 504 eventually appears and no file is downloaded
- When downloading CSV file, the download starts and continues for very close to the 30 second timeout period, then completes
    - The number of rows included varies, but there doesn't seem to ever be an incomplete row ie corrupted data

@istride @ChrisMarsh82

**Comment by cbunicef** [comment]
When trying to download XLSX I now get:

```
An error occurred.
Sorry, the page you are looking for is currently unavailable.
Please try again later.

If you are the system administrator of this resource then you should check the error log for details.

Faithfully yours, nginx.
```

When downloading CSV the download stops at 30 seconds, with around 5K of 24K rows included.

**Comment by cbunicef** [comment]
Thanks @istride. If we think the time it takes to run the query will have an impact on the API development then let's please consider that early.

**Comment by ChrisMarsh82** [comment]
Change CSV to not stream. It will only download once fully complete. This will stop users from getting incomplete data and thinking they have a complete set data

Investigate: can we give a specific error when we get a timeout error in this area?

**Comment by ChrisMarsh82** [comment]
Possible solution: downloads go to another area that users can download from once all data is collected

**Comment by istride** [comment]
It has been quite straightforward to configure threads on our sandbox instance, and I have not noticed any adverse effects so far. I have taken the liberty of configuring the South Africa instance in the same way, to see whether this issue still exists or not.

I'm glad to say that I managed to download what I think is the complete set of results (27,368 rows inc. header). The XLSX download was also successful, in that something was downloaded, however, there were only 23,601 rows in the Excel file, and I can't explain why that is.

There is another issue with the Excel download in that there is very little indication that anything is happening once the download button is clicked, which may lead a frustrated user to click again, and again, eventually occupying several worker threads with the same task simultaneously, possibly leading to downtime for the site, in general. I have no proof that this would happen, but in any case, I think it would be prudent to give more feedback to users about the download process.

**Comment by cbunicef** [comment]
Great to know that the threading has worked!

The resulting data is a strange case... when I downloaded the CSV file a few minutes ago I got 23,601 rows. When I downloaded the XLSX a few mins ago I got 23,601 as well. Based on the number of paginations shown in the Admin Panel (1180) and the rows per page (20, last page has all 20 rows), we would expect 23,601 rows including header. So I'm not sure why you got so many rows when you downloaded the CSV - but I'm also suspicious that we would have exactly 23,600 responses.

Now I am seeing that downloading the XLSX results in gateway timeout, CSV is still working.

For user feedback, the simplest solution might be spawning a new tab when the button is clicked, and the download begins there.
